packages/grid/_modules_/grid/hooks/root/useContainerProps.ts

```diff
@@ -40,124 +41,172 @@ export const useContainerProps = (windowRef: React.RefObject<HTMLDivElement>, ap
+
+      logger.debug(
+        `viewportPageSize:  ${viewportPageSize}, rzPageSize: ${rzPageSize}, viewportMaxPage: ${viewportMaxPage}`,
+      );
+      const renderingZoneHeight =
+        rzPageSize * options.rowHeight + options.rowHeight + scrollState.scrollBarSize.x;
+      const dataContainerWidth = columnsTotalWidth - scrollState.scrollBarSize.y;
+      let totalHeight =
+        (options.autoPageSize ? 1 : rowsCount / viewportPageSize) * viewportSizes.height +
+        (scrollState.hasScrollY ? scrollState.scrollBarSize.x : 0);
+
+      if (options.autoHeight) {
+        totalHeight = rowsCount * options.rowHeight + scrollState.scrollBarSize.x;
+      }
+
+      const indexes: ContainerProps = {
+        virtualRowsCount: options.autoPageSize ? viewportPageSize : rowsCount,
+        renderingZonePageSize: rzPageSize,
+        viewportPageSize,
+        totalSizes: {
+          width: columnsTotalWidth,
+          height: totalHeight || 1,
+        },
+        dataContainerSizes: {
+          width: dataContainerWidth,
+          height: totalHeight || 1,
+        },
+        renderingZone: {
+          width: dataContainerWidth,
+          height: renderingZoneHeight,
+        },
+        windowSizes: windowSizesRef.current,
+        lastPage: viewportMaxPage,
+      };
+
+      logger.debug('returning container props', indexes);
+      return indexes;
+    },
+    [
+      windowRef,
+      columnsTotalWidth,
+      options.rowHeight,
+      options.pagination,
+      options.autoPageSize,
+      options.autoHeight,
+      logger,
+    ],
+  );
+
+  const updateStateIfChanged = React.useCallback(
+    (
+      shouldUpdate: (oldState: GridState) => boolean,
+      newStateUpdate: (state: GridState) => GridState,
+    ) => {
+      let update = false;
       setGridState((state) => {
-        if (!isEqual(state.containerSizes, containerState)) {
-          state.containerSizes = containerState;
-          updated = true;
-          return { ...state };
+        update = shouldUpdate(state);
+        if (update) {
+          return newStateUpdate(state);
         }
         return state;
       });
-      if (updated) {
+      if (update) {
         forceUpdate();
       }
     },
     [forceUpdate, setGridState],
   );
 
   const refreshContainerSizes = React.useCallback(() => {
-    const containerProps = getContainerProps();
-    updateContainerState(containerProps);
-  }, [getContainerProps, updateContainerState]);
+    const rowsCount = getVirtualRowCount();
+    const scrollState = getScrollbar(rowsCount);
+
+    const viewportSize = getViewport(rowsCount, scrollState);
+    if (!viewportSize) {
+      return;
+    }
+
+    updateStateIfChanged(
+      (state) => state.scrollBar !== scrollState,
+      (state) => ({ ...state, scrollBar: scrollState }),
+    );
+
+    updateStateIfChanged(
+      (state) => state.viewportSizes !== viewportSize,
+      (state) => ({ ...state, viewportSizes: viewportSize }),
```

Comment:
Is there a way we could use the same name for the variable?

```suggestion
      (state) => ({ ...state, viewportSizes }),
```